File: QytetetRuby/pkg/QytetetRuby-0.0.1/lib/jugador.rb
#encoding: utf-8

require_relative "titulo_propiedad"
require_relative "casilla"
require_relative "especulador"

module ModeloQytetet
  class Jugador
    attr_accessor :cartaLibertad, :casillaActual, :encarcelado, :nombre, :propiedades, :saldo
    
    def initialize(nombre, saldo = 7500, encarcelado = false, propiedades = Array.new, carta = nil, casilla = nil)
      @nombre = nombre
      @saldo = saldo
      @encarcelado = encarcelado
      @propiedades = propiedades
      @cartaLibertad = carta
      @casillaActual = casilla
    end
    
    def <=>(otroJugador)            
      otroJugador.obtenerCapital <=> obtenerCapital     
    end
    
    def cancelarHipoteca(titulo)
      cancelar = false
      cantidadRecibida = titulo.calcularCosteCancelar
      
      if @saldo >= cantidadRecibida
        modificarSaldo(-cantidadRecibida)
        cancelar = titulo.cancelarHipoteca
      end
      
      return cancelar
    end
    
    def comprarTituloPropiedad
      comprado = false
      
      costeCompra = @casillaActual.coste
      
      if costeCompra < @saldo
        @casillaActual.asignarPropietarioJugador(self)
        comprado = true
        @propiedades << @casillaActual.titulo
        modificarSaldo(-costeCompra)
      end
      
      return comprado
    end
    
    def cuantasCasasHotelesTengo
      casas_hoteles = 0
      
      propiedades.each{|x|
        casas_hoteles = x.numCasas + 1
        casas_hoteles = x.numHoteles + 1
      }
      
      return casas_hoteles
    end
    
    def deboPagarAlquiler
      deboPagar = false
      titulo = @casillaActual.titulo
      
      if !esDeMiPropiedad(titulo) && titulo.tengoPropietario && !titulo.propietarioEncarcelado && !titulo.hipotecada
        deboPagar = true
      end
      
      return deboPagar
    end
    
    def deboIrACarcel
      deboIr = true
      
      if tengoCartaLibertad
        deboIr = false
      end
      
      return deboIr
    end
    
    def devolverCartaLibertad
      auxiliar = @cartaLibertad
      @cartaLibertad = nil
      
      return auxiliar
    end
    
    def edificarCasa(titulo)
      edificada = false
      
      if puedoEdificarCasa(titulo)
        costeEdificarCasa = titulo.precioEdificar
        
        if tengoSaldo(costeEdificarCasa)
          titulo.edificarCasa
          modificarSaldo(-costeEdificarCasa)
          edificada = true
        end
      end
      
      return edificada;
    end
    
    def edificarHotel(titulo)
      edificada = false
      
      if puedoEdificarHotel(titulo)
        costeEdificarHotel = titulo.precioEdificar
        
        if tengoSaldo(costeEdificarHotel)
          titulo.edificarHotel
          modificarSaldo(-costeEdificarHotel)
          edificada = true
        end
      end
      
      return edificada;
    end
    
    def eliminarDeMisPropiedades(titulo)
      @propiedades.delete(titulo)
      titulo.propietario = nil
    end
    
    def esDeMiPropiedad(titulo)
      contiene = false
 
      @propiedades.each {|x| 
        if x.nombre == titulo.nombre then contiene = true end}
      
      return contiene
    end
    
    def estoyEnCalleLibre
      raise NotImplementedError
    end
        
    def hipotecarPropiedad(titulo)
      costeHipoteca = titulo.hipotecar
      modificarSaldo(costeHipoteca)
    end
    
    def irACarcel(casilla)
      @casillaActual = casilla
      @encarcelado = true
    end
    
    def modificarSaldo(cantidad)
      @saldo = @saldo + cantidad
      
      return @saldo
    end
    
    def obtenerCapital
      capital = 0
      
      @propiedades.each{|x| capital = capital + (x.precioCompra + x.numCasas+ x.numHoteles)*x.precioEdificar
          if x.hipotecada
            capital = capital - x.hipotecaBase
          end}
      
      capital = capital + @saldo
      
      return capital
    end
    
    def obtenerPropiedades(hipotecada)
      obtener_propiedades = Array.new
      
      @propiedades.each{|x|
        if x.hipotecada == hipotecada 
          obtener_propiedades << x
        end}
      
      return obtener_propiedades
    end
    
    def pagarAlquiler
      costeAlquiler = @casillaActual.pagarAlquiler
      
      puts "\nEsta casilla tiene dueño, pagas el alquiler de " + costeAlquiler.to_s + " euros"
      
      modificarSaldo(-costeAlquiler)
    end
    
    def pagarImpuesto
      @saldo = @saldo - @casillaActual.coste
    end
    
    def pagarLibertad(cantidad)
      tengoSaldo = tengoSaldo(cantidad)
      
      if tengoSaldo
        @encarcelado = false
        modificarSaldo(-cantidad)
      end
    end
    
    def tengoCartaLibertad
      tengocartaLibertad = false
      
      if @cartaLibertad!=nil
        tengocartaLibertad = true
      end
      
      return tengocartaLibertad
    end
    
    def tengoSaldo(cantidad)
      if @saldo > cantidad
        return true
      else
        return false
      end
    end
    
    def venderPropiedad(casilla)
      titulo = casilla.titulo
      eliminarDeMisPropiedades(titulo)
      precioVenta = titulo.calcularPrecioVenta
      modificarSaldo(precioVenta)
    end
    
    def convertirme(fianza)
      especulador = Especulador.new(self,fianza)
      
      return especulador
    end
    
    def puedoEdificarCasa(titulo) 
      puedo_edificar = false
      
      if titulo.numCasas < 4
        puedo_edificar = true
      end
      
      return puedo_edificar
    end
    
    def puedoEdificarHotel(titulo)
      puedo_edificar = false
      
      if titulo.numHoteles < 4 && titulo.numCasas == 4
        puedo_edificar = true
      end
      
      return puedo_edificar
    end
    
    def to_s
      "Nombre " +@nombre+" \n Saldo #{@saldo} \n CartaLibertad #{@cartaLibertad} \n CasillaActual #{@casillaActual} \n Encarcelado #{@encarcelado} \n Propiedades #{@propiedades}"
    end
    
    protected  :deboIrACarcel, :pagarImpuesto, :puedoEdificarCasa,
               :puedoEdificarHotel, :tengoSaldo
    private :eliminarDeMisPropiedades, :esDeMiPropiedad
  end
end

File: QytetetJava/src/modeloqytetet/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import java.util.ArrayList;

/**
 *
 * @author Juan Carlos Hermoso Quesada
 */
public class Jugador implements Comparable{
    private boolean encarcelado = false;
    private String nombre;
    private int saldo = 7500;
    private Sorpresa cartaLibertad;
    private Casilla casillaActual;
    private ArrayList<TituloPropiedad> propiedades = new ArrayList();
    
    public Jugador(String nombre) {
        this.nombre = nombre;
    }
    
    public Jugador(Jugador otroJugador) {
        this.nombre = otroJugador.nombre;
        this.saldo = otroJugador.saldo;
        this.encarcelado = otroJugador.encarcelado;
        this.cartaLibertad = otroJugador.cartaLibertad;
        this.casillaActual = otroJugador.casillaActual;
        this.propiedades = new ArrayList<>(otroJugador.getPropiedades());
    } 
    
    @Override
    public int compareTo(Object otroJugador) {
        int otroCapital = ((Jugador) otroJugador).obtenerCapital();
        return otroCapital - obtenerCapital();
    }
        
    boolean cancelarHipoteca(TituloPropiedad titulo) {
        boolean cancelar = false;
       
        int cantidadRecibida = titulo.calcularCosteCancelar();
        
        if (this.saldo >= cantidadRecibida) {
            modificarSaldo(-cantidadRecibida);
            cancelar = titulo.cancelarHipoteca();
        }
       
        return cancelar;
    }
    
    boolean comprarTituloPropiedad() {
        boolean comprado = false;
        
        int costeCompra = casillaActual.getCoste();
        
        Calle calle = (Calle)casillaActual;
        
        if (costeCompra < saldo) {
            calle.asignarPropietarioJugador(this);
            comprado = true;
            propiedades.add(calle.getTitulo());
            modificarSaldo(-costeCompra);
        }
        
        return comprado;
    }
    
    int cuantasCasasHotelesTengo() {
        int casas_hoteles = 0;
        
        for (int i = 0; i < propiedades.size(); i++) {
            casas_hoteles += propiedades.get(i).getNumCasas();
            casas_hoteles += propiedades.get(i).getNumHoteles();
        }
        
        return casas_hoteles;
    }
        
    boolean deboPagarAlquiler() {
        boolean deboPagar = false;
        
        TituloPropiedad titulo = casillaActual.getTitulo();
        
        if (!esDeMiPropiedad(titulo) && titulo.tengoPropietario() && !titulo.propietarioEncarcelado()
           && !titulo.isHipotecada()) {
            deboPagar = true;
        } 
        
        return deboPagar;
    }
    
    Sorpresa devolverCartaLibertad() {
        Sorpresa auxiliar;
        auxiliar = cartaLibertad;
        cartaLibertad = null;
        
        return auxiliar;        
    }
    
    boolean edificarCasa(TituloPropiedad titulo) {
        boolean edificada = false;
                
        if (puedoEdificarCasa(titulo)) {
            int costeEdificarCasa = titulo.getPrecioEdificar();
            
            if (tengoSaldo(costeEdificarCasa)) {
                titulo.edificarCasa();
                modificarSaldo(-costeEdificarCasa);
                edificada = true;
            }
        }
        
        return edificada;
    }
    
    boolean edificarHotel(TituloPropiedad titulo) {
        boolean edificada = false;
        
        if (puedoEdificarHotel(titulo)) {
            int costeEdificarHotel = titulo.getPrecioEdificar();
            
            if (tengoSaldo(costeEdificarHotel)) {
                titulo.edificarHotel();
                modificarSaldo(-costeEdificarHotel);
                edificada = true;
            }
        }
        
        return edificada;
    }
    
    private void eliminarDeMisPropiedades(TituloPropiedad titulo) {
        propiedades.remove(titulo);
        titulo.setPropietario(null);
    }
    
    private boolean esDeMiPropiedad(TituloPropiedad titulo) {
        boolean contiene = false;
        
        for (int i = 0; i < propiedades.size(); i++) {
            if (propiedades.get(i).getNombre() == titulo.getNombre()) {
                contiene = true;
            }
        }
        
        return contiene;
    }
    
    boolean estoyEnCalleLibre() {
        throw new UnsupportedOperationException("Sin implementar");
    }
    
    Sorpresa getCartaLibertad() {
        return cartaLibertad;
    }
    
    public Casilla getCasillaActual() {
        return casillaActual;
    }
    
    boolean getEncarcelado() {
        return encarcelado;
    }
    
    public String getNombre() {
        return nombre;
    }
    
    public ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }
    
    public int getSaldo() {
        return saldo;
    }
    
    void hipotecarPropiedad(TituloPropiedad titulo) {
        int costeHipoteca = titulo.hipotecar();
        this.modificarSaldo(costeHipoteca);
    }
    
    void irACarcel(Casilla casilla) {
        setCasillaActual(casilla);
        setEncarcelado(true);
    }
    
    int modificarSaldo(int cantidad) {
        saldo += cantidad;
        
        return saldo;
    }
    
    int obtenerCapital() {
        int capital = 0;
        
        for (int i = 0; i < propiedades.size(); i++) {
            capital += (propiedades.get(i).getNumCasas() + propiedades.get(i).getNumHoteles())*propiedades.get(i).getPrecioEdificar() + propiedades.get(i).getPrecioCompra();
            
            if (propiedades.get(i).isHipotecada()) {
                capital -= propiedades.get(i).getHipotecaBase();
            }
        }
        return capital+saldo;
    }
    
    ArrayList<TituloPropiedad> obtenerPropiedades(boolean hipotecada) {
        ArrayList<TituloPropiedad> obtener_propiedades = new ArrayList();
        
        for (int i = 0; i < propiedades.size(); i++) {
            if (propiedades.get(i).isHipotecada() == hipotecada) {
                obtener_propiedades.add(propiedades.get(i));
            }
        }
        
        return obtener_propiedades;
    }
    
    void pagarAlquiler() {
        Calle calle = (Calle)casillaActual;
        
        int costeAlquiler = calle.pagarAlquiler();
        
        System.out.println("\nEsta casilla tiene dueño, pagas el alquiler de "+ costeAlquiler + " euros.\n");
        
        modificarSaldo(-costeAlquiler);
    }   
    
    protected void pagarImpuesto() {
        saldo -= casillaActual.getCoste();
    }
    
    void pagarLibertad(int cantidad) {
        boolean tengoSaldo = tengoSaldo(cantidad);
        
        if (tengoSaldo) {
            setEncarcelado(false);
            modificarSaldo(-cantidad);
        }
    }
    
    void setCartaLibertad(Sorpresa carta) {
        this.cartaLibertad = carta;
    }
    
    void setCasillaActual(Casilla casilla) {
        this.casillaActual = casilla;
    }
    
    void setEncarcelado(boolean encarcelado) {
        this.encarcelado = encarcelado;
    }
    
    boolean tengoCartaLibertad() {
        boolean tengocartalibertad = false;
        
        if (cartaLibertad!=null) {
            tengocartalibertad = true;
        }
        
        return tengocartalibertad;
    }
    
    protected boolean tengoSaldo(int cantidad) {
        return saldo > cantidad;
    }
    
    void venderPropiedad(Casilla casilla) {
        TituloPropiedad titulo = casilla.getTitulo();
        eliminarDeMisPropiedades(titulo);
        int precioVenta = titulo.calcularPrecioVenta();
        modificarSaldo(precioVenta);
    }
    
    protected Especulador convertirme(int fianza) {
        
        Especulador especulador = new Especulador(fianza,this);
        
        return especulador;
    }
    
    @Override
    public String toString() {
        return "Jugador " + nombre + "\n      Saldo= " + saldo + "\n      Encarcelado= " + encarcelado + "\n      Capital= "+ this.obtenerCapital() +"\n      CartaLibertad= " + cartaLibertad + "\n\n      CasillaActual= " + casillaActual + "\n\n      PROPIEDADES\n " + propiedades;
    }
    
    protected boolean deboIrACarcel() {
        boolean deboIrACarcel;
        
        if (tengoCartaLibertad()) {
            deboIrACarcel = false;
        }
        else {
            deboIrACarcel = true;
        }
        
        return deboIrACarcel;
    }
    
    protected boolean puedoEdificarCasa(TituloPropiedad titulo) {
        boolean puedo_edificar = false;
        
        if (titulo.getNumCasas() < 4) {
            puedo_edificar = true;
        }
        
        return puedo_edificar;
    }
    
    protected boolean puedoEdificarHotel(TituloPropiedad titulo) {
        boolean puedo_edificar = false;
        
        if (titulo.getNumHoteles() < 4 && titulo.getNumCasas() == 4) {
            puedo_edificar = true;
        }
        
        return puedo_edificar;    
    }
    
    
       
    
}
